pmlc/util/strides.cc

```diff
@@ -0,0 +1,91 @@
+// Copyright 2020 Intel Corporation
+
+#include "pmlc/util/strides.h"
+
+#include <vector>
+
+#include "mlir/Analysis/AffineStructures.h"
+#include "pmlc/util/logging.h"
+
+namespace pmlc::util {
+
+StrideArray::StrideArray(unsigned numDims, int64_t offset)
+    : offset(offset), strides(numDims) {}
+
+StrideArray &StrideArray::operator*=(int64_t factor) {
```

Comment:
Do you want to support factor < 0?

Reply:
Yes, technically negative stride memrefs would cause such an effect, and I see no reason to disallow them (and upstream MLIR does, although given that we currently only use 'natural' layout, they will never occur).